File: Python-Bank-Project/blockchain_integration.py
#!/usr/bin/env python3
"""
Blockchain Integration Module for Banking System
Handles recording donations on the blockchain smart contract
"""

import json
import time
import requests
from web3 import Web3

# Configuration
GANACHE_URL = "http://127.0.0.1:7545"  # Ganache RPC URL
CONTRACT_ADDRESS = "0x9fC0c4B491bC255f1d1486aD586d404b425afD8F"  # From contract-address.json
CHAIN_ID = 1337  # Ganache chain ID

# Contract ABI (updated from contract-abi.js)
CONTRACT_ABI = [
    {
        "inputs": [],
        "stateMutability": "nonpayable",
        "type": "constructor"
    },
    {
        "anonymous": False,
        "inputs": [
            {"indexed": True, "internalType": "uint256", "name": "transactionId", "type": "uint256"},
            {"indexed": False, "internalType": "string", "name": "ngoId", "type": "string"},
            {"indexed": False, "internalType": "string", "name": "donorId", "type": "string"},
            {"indexed": False, "internalType": "string", "name": "cause", "type": "string"},
            {"indexed": False, "internalType": "uint256", "name": "amount", "type": "uint256"},
            {"indexed": False, "internalType": "uint256", "name": "timestamp", "type": "uint256"}
        ],
        "name": "DonationReceived",
        "type": "event"
    },
    {
        "anonymous": False,
        "inputs": [
            {"indexed": True, "internalType": "uint256", "name": "transactionId", "type": "uint256"},
            {"indexed": False, "internalType": "string", "name": "ngoId", "type": "string"},
            {"indexed": False, "internalType": "string", "name": "receiverId", "type": "string"},
            {"indexed": False, "internalType": "string", "name": "cause", "type": "string"},
            {"indexed": False, "internalType": "uint256", "name": "amount", "type": "uint256"},
            {"indexed": False, "internalType": "uint256", "name": "timestamp", "type": "uint256"},
            {"indexed": False, "internalType": "bytes32", "name": "verificationHash", "type": "bytes32"}
        ],
        "name": "FundsSpent",
        "type": "event"
    },
    {
        "inputs": [],
        "name": "owner",
        "outputs": [{"internalType": "address", "name": "", "type": "address"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [
            {"internalType": "string", "name": "ngoId", "type": "string"},
            {"internalType": "string", "name": "donorId", "type": "string"},
            {"internalType": "string", "name": "cause", "type": "string"},
            {"internalType": "uint256", "name": "amount", "type": "uint256"},
            {"internalType": "uint256", "name": "timestamp", "type": "uint256"}
        ],
        "name": "recordDonation",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "inputs": [
            {"internalType": "string", "name": "ngoId", "type": "string"},
            {"internalType": "string", "name": "receiverId", "type": "string"},
            {"internalType": "string", "name": "cause", "type": "string"},
            {"internalType": "uint256", "name": "amount", "type": "uint256"},
            {"internalType": "uint256", "name": "timestamp", "type": "uint256"},
            {"internalType": "bytes32", "name": "verificationHash", "type": "bytes32"}
        ],
        "name": "recordSpending",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "string", "name": "ngoId", "type": "string"}],
        "name": "getNgoBalance",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "totalDonations",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "getIncomingDonationsCount",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "getOutgoingTransactionsCount",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    }
]

class BlockchainIntegration:
    def __init__(self):
        self.web3 = None
        self.contract = None
        self.account = None
        self.is_connected = False
        
    def connect(self):
        """Connect to Ganache blockchain"""
        try:
            # Connect to Ganache
            self.web3 = Web3(Web3.HTTPProvider(GANACHE_URL))
            
            # Check connection
            if not self.web3.is_connected():
                raise Exception("Failed to connect to Ganache")
            
            # Get accounts (use first account as default)
            accounts = self.web3.eth.accounts
            if not accounts:
                raise Exception("No accounts found in Ganache")
            
            # Use the specified Ganache account
            self.account = "0x35b6cdc6F2a0990d38d232eEe6007846B531d5a0"
            
            # Load contract
            self.contract = self.web3.eth.contract(
                address=Web3.to_checksum_address(CONTRACT_ADDRESS),
                abi=CONTRACT_ABI
            )
            
            self.is_connected = True
            print(f"✅ Connected to blockchain successfully")
            print(f"   Account: {self.account}")
            print(f"   Contract: {CONTRACT_ADDRESS}")
            
            return True
            
        except Exception as e:
            print(f"❌ Blockchain connection failed: {e}")
            self.is_connected = False
            return False
    
    def record_donation_on_blockchain(self, ngo_account, donor_id, cause, amount):
        """
        Record a donation on the blockchain smart contract
        
        Args:
            ngo_account (str): NGO's bank account number (used as NGO ID)
            donor_id (str): Donor's identification (MongoDB ObjectID from donation system)
            cause (str): Purpose of the donation (education, health, etc.)
            amount (int): Donation amount in rupees
            
        Returns:
            dict: Result with success status and transaction details
        """
        if not self.is_connected:
            if not self.connect():
                return {
                    'success': False,
                    'error': 'Blockchain connection failed',
                    'tx_hash': None,
                    'blockchain_tx_id': None
                }
        
        try:
            # Prepare transaction parameters
            ngo_id = f"NGO_{ngo_account}"  # Convert bank account to NGO ID
            timestamp = int(time.time())
            
            print(f"🔗 Recording donation on blockchain:")
            print(f"   NGO ID: {ngo_id}")
            print(f"   Donor ID: {donor_id}")
            print(f"   Cause: {cause}")
            print(f"   Amount: ₹{amount}")
            
            # For Ganache (unlocked accounts), we can send transaction directly
            tx_hash = self.contract.functions.recordDonation(
                ngo_id,
                donor_id,
                cause or "general",  # Default to "general" if no cause provided
                amount,
                timestamp
            ).transact({
                'from': self.account,
                'gas': 500000
            })
            
            print(f"📤 Transaction sent: {tx_hash.hex()}")
            
            # Wait for transaction receipt
            tx_receipt = self.web3.eth.wait_for_transaction_receipt(tx_hash, timeout=30)
            
            if tx_receipt.status == 1:
                # Parse logs to get the blockchain transaction ID
                blockchain_tx_id = None
                if tx_receipt.logs:
                    try:
                        # Parse the DonationReceived event
                        donation_event = self.contract.events.DonationReceived().process_log(tx_receipt.logs[0])
                        blockchain_tx_id = donation_event['args']['transactionId']
                    except Exception as log_error:
                        print(f"⚠️ Could not parse transaction ID from logs: {log_error}")
                
                print(f"✅ Donation recorded on blockchain successfully!")
                print(f"   Transaction Hash: {tx_hash.hex()}")
                print(f"   Blockchain Transaction ID: {blockchain_tx_id}")
                
                return {
                    'success': True,
                    'tx_hash': tx_hash.hex(),
                    'blockchain_tx_id': blockchain_tx_id,
                    'block_number': tx_receipt.blockNumber,
                    'gas_used': tx_receipt.gasUsed
                }
            else:
                return {
                    'success': False,
                    'error': 'Transaction failed on blockchain',
                    'tx_hash': tx_hash.hex(),
                    'blockchain_tx_id': None
                }
                
        except Exception as e:
            print(f"❌ Error recording donation on blockchain: {e}")
            return {
                'success': False,
                'error': str(e),
                'tx_hash': None,
                'blockchain_tx_id': None
            }
    
    def record_spending_on_blockchain(self, ngo_account, receiver_id, cause, amount):
        """
        Record spending/withdrawal on the blockchain smart contract
        
        Args:
            ngo_account (str): NGO's bank account number (used as NGO ID)
            receiver_id (str): Receiver of the funds (for withdrawals, this could be the account holder)
            cause (str): Purpose of the spending (Cash Withdrawal, Fund Transfer, etc.)
            amount (int): Amount spent in rupees
            
        Returns:
            dict: Result with success status and transaction details
        """
        if not self.is_connected:
            if not self.connect():
                return {
                    'success': False,
                    'error': 'Blockchain connection failed',
                    'tx_hash': None,
                    'blockchain_tx_id': None
                }
        
        try:
            # Prepare transaction parameters
            ngo_id = f"NGO_{ngo_account}"
            timestamp = int(time.time())
            verification_hash = self.web3.keccak(text=f"spending_{ngo_account}_{timestamp}")
            
            print(f"🔗 Recording spending on blockchain:")
            print(f"   NGO ID: {ngo_id}")
            print(f"   Receiver ID: {receiver_id}")
            print(f"   Cause: {cause}")
            print(f"   Amount: ₹{amount}")
            
            # Record spending on blockchain using recordSpending function
            tx_hash = self.contract.functions.recordSpending(
                ngo_id,
                receiver_id,
                cause or "general_spending",
                amount,
                timestamp,
                verification_hash
            ).transact({
                'from': self.account,
                'gas': 500000
            })
            
            print(f"📤 Spending transaction sent: {tx_hash.hex()}")
            
            # Wait for transaction receipt
            tx_receipt = self.web3.eth.wait_for_transaction_receipt(tx_hash, timeout=30)
            
            if tx_receipt.status == 1:
                # Parse logs to get the blockchain transaction ID
                blockchain_tx_id = None
                if tx_receipt.logs:
                    try:
                        # Parse the FundsSpent event
                        funds_spent_event = self.contract.events.FundsSpent().process_log(tx_receipt.logs[0])
                        blockchain_tx_id = funds_spent_event['args']['transactionId']
                    except Exception as log_error:
                        print(f"⚠️ Could not parse transaction ID from logs: {log_error}")
                
                print(f"✅ Spending recorded on blockchain successfully!")
                print(f"   Transaction Hash: {tx_hash.hex()}")
                print(f"   Blockchain Transaction ID: {blockchain_tx_id}")
                
                return {
                    'success': True,
                    'tx_hash': tx_hash.hex(),
                    'blockchain_tx_id': blockchain_tx_id,
                    'block_number': tx_receipt.blockNumber,
                    'gas_used': tx_receipt.gasUsed
                }
            else:
                return {
                    'success': False,
                    'error': 'Transaction failed on blockchain',
                    'tx_hash': tx_hash.hex(),
                    'blockchain_tx_id': None
                }
                
        except Exception as e:
            print(f"❌ Error recording spending on blockchain: {e}")
            return {
                'success': False,
                'error': str(e),
                'tx_hash': None,
                'blockchain_tx_id': None
            }

    def get_ngo_balance_from_blockchain(self, ngo_account):
        """Get NGO balance from blockchain"""
        if not self.is_connected:
            if not self.connect():
                return None
        
        try:
            ngo_id = f"NGO_{ngo_account}"
            balance = self.contract.functions.getNgoBalance(ngo_id).call()
            return balance
        except Exception as e:
            print(f"❌ Error getting NGO balance from blockchain: {e}")
            return None
    
    def _get_private_key(self):
        """
        Get private key for signing transactions
        In production, use proper key management (env vars, key stores, etc.)
        For Ganache development, use the specified test key
        """
        # Your specific Ganache account private key
        # This is a test key for development only - NEVER use in production
        if self.account == "0x35b6cdc6F2a0990d38d232eEe6007846B531d5a0":
            return "0x69c3cf937091d5c71fe45ca0e738e5c54c96ddc233e8b"
        
        # Fallback - this shouldn't happen with the fixed account
        print(f"⚠️ Unknown account: {self.account}")
        return None
    
    def get_blockchain_status(self):
        """Get blockchain connection status and basic info"""
        if not self.is_connected:
            return {
                'connected': False,
                'error': 'Not connected to blockchain'
            }
        
        try:
            latest_block = self.web3.eth.block_number
            total_donations = self.contract.functions.totalDonations().call()
            
            return {
                'connected': True,
                'latest_block': latest_block,
                'contract_address': CONTRACT_ADDRESS,
                'account': self.account,
                'total_donations_on_chain': total_donations,
                'chain_id': CHAIN_ID
            }
        except Exception as e:
            return {
                'connected': False,
                'error': str(e)
            }

# Global instance
blockchain = BlockchainIntegration()

# Test function
if __name__ == "__main__":
    # Test the blockchain integration
    print("🧪 Testing Blockchain Integration...")
    
    # Connect to blockchain
    if blockchain.connect():
        print("\n📊 Blockchain Status:")
        status = blockchain.get_blockchain_status()
        for key, value in status.items():
            print(f"   {key}: {value}")
        
        # Test recording a donation
        print("\n🎯 Testing donation recording...")
        result = blockchain.record_donation_on_blockchain(
            ngo_account="12345678",
            donor_id="TEST_DONOR_001", 
            cause="education",
            amount=500
        )
        
        print(f"\n📋 Result: {json.dumps(result, indent=2)}")
    else:
        print("❌ Failed to connect to blockchain")

File: backend/utils/blockchainIntegration.js
// /**
//  * Blockchain integration utility for recording spending transactions
//  * This is a placeholder that can be replaced with actual blockchain calls
//  */

// import { ethers } from 'ethers';
// import { BLOCKCHAIN_CONFIG, CONTRACT_ABI } from '../config/blockchain.config.js';

// class BlockchainIntegration {
//   constructor() {
//     this.isInitialized = false;
//     this.provider = null;
//     this.signer = null;
//     this.contract = null;
//     this.connectionAttempts = 0;
//   }

//   async initialize() {
//     try {
//       console.log('🔗 Initializing blockchain connection...');
//       console.log(`   RPC URL: ${BLOCKCHAIN_CONFIG.RPC_URL}`);
//       console.log(`   Contract: ${BLOCKCHAIN_CONFIG.CONTRACT_ADDRESS}`);
      
//       // Create provider
//       this.provider = new ethers.JsonRpcProvider(BLOCKCHAIN_CONFIG.RPC_URL, {
//         chainId: 1337,
//         name: 'Ganache'
//       });
      
//       // Test connection
//       const blockNumber = await this.provider.getBlockNumber();
//       console.log(`   Connected! Current block: ${blockNumber}`);
      
//       // Create signer from private key
//       this.signer = new ethers.Wallet(BLOCKCHAIN_CONFIG.PRIVATE_KEY, this.provider);
//       const signerAddress = await this.signer.getAddress();
//       console.log(`   Signer address: ${signerAddress}`);
      
//       // Create contract instance
//       this.contract = new ethers.Contract(
//         BLOCKCHAIN_CONFIG.CONTRACT_ADDRESS,
//         CONTRACT_ABI,
//         this.signer
//       );
      
//       // Verify contract is accessible
//       const owner = await this.contract.owner();
//       console.log(`   Contract owner: ${owner}`);
      
//       if (owner.toLowerCase() !== signerAddress.toLowerCase()) {
//         console.warn('⚠️ Warning: Signer is not contract owner. recordSpending will fail!');
//         console.warn(`   Expected: ${owner}`);
//         console.warn(`   Got: ${signerAddress}`);
//       }
      
//       this.isInitialized = true;
//       this.connectionAttempts = 0;
//       return true;
      
//     } catch (error) {
//       console.error('❌ Failed to initialize blockchain:', error.message);
//       this.isInitialized = false;
//       this.connectionAttempts++;
      
//       // If too many failures, throw error
//       if (this.connectionAttempts >= 3) {
//         throw new Error(`Blockchain connection failed after ${this.connectionAttempts} attempts: ${error.message}`);
//       }
      
//       return false;
//     }
//   }

//   async recordSpending({ ngo_account, receiver_id, cause, amount, verification_hash }) {
//     try {
//       // Ensure we're connected
//       if (!this.isInitialized) {
//         const initialized = await this.initialize();
//         if (!initialized) {
//           throw new Error('Failed to initialize blockchain connection');
//         }
//       }

//       console.log(`🔗 Recording spending on blockchain:`);
//       console.log(`   NGO Account: ${ngo_account}`);
//       console.log(`   Receiver: ${receiver_id}`);
//       console.log(`   Cause: ${cause}`);
//       console.log(`   Amount: ₹${amount}`);
//       console.log(`   Verification Hash: ${verification_hash}`);

//       // Convert verification hash to bytes32 format
//       let hashBytes32;
//       if (!verification_hash || verification_hash === '0x0') {
//         // Use zero bytes32 for missing documents
//         hashBytes32 = ethers.ZeroHash;
//       } else {
//         // Ensure hash is proper bytes32 format
//         hashBytes32 = verification_hash.startsWith('0x') ? verification_hash : `0x${verification_hash}`;
//         // Pad to 32 bytes if needed
//         if (hashBytes32.length < 66) {
//           hashBytes32 = ethers.zeroPadValue(hashBytes32, 32);
//         }
//       }

//       // Record spending on blockchain
//       // Note: The contract expects (ngoId, receiverId, cause, amount, timestamp, verificationHash)
//       const timestamp = Math.floor(Date.now() / 1000); // Current time in seconds
      
//       console.log(`📤 Sending transaction to blockchain...`);
//       const tx = await this.contract.recordSpending(
//         ngo_account,
//         receiver_id,
//         cause,
//         amount,
//         timestamp,
//         hashBytes32,
//         {
//           gasLimit: BLOCKCHAIN_CONFIG.GAS_LIMIT
//         }
//       );
      
//       console.log(`⏳ Transaction sent: ${tx.hash}`);
//       console.log(`   Waiting for confirmation...`);
      
//       // Wait for transaction to be mined
//       const receipt = await tx.wait();
      
//       console.log(`✅ Transaction confirmed in block ${receipt.blockNumber}`);
//       console.log(`   Gas used: ${receipt.gasUsed.toString()}`);
      
//       return {
//         success: true,
//         tx_id: tx.hash,
//         blockchain_tx_id: tx.hash,
//         block_number: receipt.blockNumber,
//         gas_used: receipt.gasUsed.toString(),
//         verification_hash: hashBytes32,
//         timestamp: timestamp
//       };

//     } catch (error) {
//       console.error('❌ Blockchain recording failed:', error);
      
//       // Check if it's a connection error
//       if (error.code === 'NETWORK_ERROR' || error.code === 'SERVER_ERROR') {
//         console.log('🔄 Connection lost, resetting initialization...');
//         this.isInitialized = false;
//       }
      
//       return {
//         success: false,
//         error: error.message,
//         reason: error.reason || 'Unknown error'
//       };
//     }
//   }

//   async recordDonation({ ngo_account, donor_id, cause, amount }) {
//     try {
//       if (!this.isInitialized) {
//         const initialized = await this.initialize();
//         if (!initialized) {
//           throw new Error('Failed to initialize blockchain connection');
//         }
//       }

//       console.log(`🔗 Recording donation on blockchain:`);
//       console.log(`   NGO Account: ${ngo_account}`);
//       console.log(`   Donor: ${donor_id}`);
//       console.log(`   Cause: ${cause}`);
//       console.log(`   Amount: ₹${amount}`);

//       const timestamp = Math.floor(Date.now() / 1000);
      
//       const tx = await this.contract.recordDonation(
//         ngo_account,
//         donor_id,
//         cause,
//         amount,
//         timestamp,
//         {
//           gasLimit: BLOCKCHAIN_CONFIG.GAS_LIMIT
//         }
//       );
      
//       console.log(`⏳ Transaction sent: ${tx.hash}`);
//       const receipt = await tx.wait();
      
//       console.log(`✅ Donation recorded in block ${receipt.blockNumber}`);
      
//       return {
//         success: true,
//         tx_id: tx.hash,
//         blockchain_tx_id: tx.hash,
//         block_number: receipt.blockNumber,
//         gas_used: receipt.gasUsed.toString(),
//         timestamp: timestamp
//       };

//     } catch (error) {
//       console.error('❌ Failed to record donation:', error);
      
//       if (error.code === 'NETWORK_ERROR' || error.code === 'SERVER_ERROR') {
//         this.isInitialized = false;
//       }
      
//       return {
//         success: false,
//         error: error.message,
//         reason: error.reason || 'Unknown error'
//       };
//     }
//   }

//   async getTransactionStatus(txId) {
//     try {
//       if (!this.isInitialized) {
//         await this.initialize();
//       }

//       console.log(`🔍 Checking blockchain transaction status: ${txId}`);
      
//       const receipt = await this.provider.getTransactionReceipt(txId);
      
//       if (!receipt) {
//         return {
//           success: true,
//           status: 'pending',
//           confirmations: 0
//         };
//       }
      
//       const currentBlock = await this.provider.getBlockNumber();
//       const confirmations = currentBlock - receipt.blockNumber;
      
//       return {
//         success: true,
//         status: receipt.status === 1 ? 'confirmed' : 'failed',
//         confirmations: confirmations,
//         block_number: receipt.blockNumber,
//         gas_used: receipt.gasUsed.toString()
//       };

//     } catch (error) {
//       console.error('❌ Failed to check transaction status:', error);
//       return {
//         success: false,
//         error: error.message
//       };
//     }
//   }

//   async getNgoBalance(ngoAccount) {
//     try {
//       if (!this.isInitialized) {
//         await this.initialize();
//       }

//       console.log(`💰 Getting NGO balance for account: ${ngoAccount}`);
      
//       const balance = await this.contract.getNgoBalance(ngoAccount);
      
//       console.log(`   Balance: ₹${balance.toString()}`);
      
//       return {
//         success: true,
//         balance: balance.toString(),
//         account: ngoAccount
//       };

//     } catch (error) {
//       console.error('❌ Failed to get NGO balance:', error);
//       return {
//         success: false,
//         error: error.message
//       };
//     }
//   }

//   /**
//    * Test blockchain connection
//    */
//   async testConnection() {
//     try {
//       if (!this.isInitialized) {
//         await this.initialize();
//       }
      
//       const blockNumber = await this.provider.getBlockNumber();
//       const owner = await this.contract.owner();
      
//       return {
//         success: true,
//         blockNumber: blockNumber,
//         contractOwner: owner,
//         connected: true
//       };
//     } catch (error) {
//       return {
//         success: false,
//         error: error.message,
//         connected: false
//       };
//     }
//   }
// }

// // Export singleton instance
// const blockchainIntegration = new BlockchainIntegration();

// export default blockchainIntegration;